peps/pep-0750.rst

```diff
@@ -0,0 +1,894 @@
+
+Unfortunately, this rewrite doesn't work if using the usual lambda wrapping to
+implement interpolations, namely ``lambda: tag``. When the interpolations are
+evaluated by the tag function, it will result in ``NameError: name 'tag' is not
+defined``. The root cause of this name error is that ``lambda: tag`` uses function scope,
+and it's therefore not able to use the class definition where ``tag`` is
+defined.
+
+Desugaring how the tag string could be evaluated will result in the same
+``NameError`` even using f-strings; the lambda wrapping here also uses function
+scoping:
+
+.. code-block:: python
+
+    class CaptionConfig:
+        tag = 'b'
+        figure = f'<{(lambda: tag)()}>Figure</{(lambda: tag)()}>'
+
+For tag strings, getting such a ``NameError`` would be surprising. It would also
+be a rough edge in using tag strings in this specific case of working with class
+variables. After all, tag strings are supposed to support a superset of the
+capabilities of f-strings.
+
+The solution is to use annotation scope for tag string interpolations. While the
+name "annotation scope" suggests it's only about annotations, it solves this
+problem by lexically resolving names in the class definition, such as ``tag``,
+unlike function scope.
+
+.. note::
+
+    The use of annotation scope means it's not possible to fully desugar
+    interpolations into Python code. Instead it's as if one is writing
+    ``interpolation_lambda: tag``, not ``lambda: tag``, where a hypothetical
+    ``interpolation_lambda`` keyword variant uses annotation scope instead of
+    the standard function scope.
+
+    This is more or less how the reference implementation implements this
+    concept (but without creating a new keyword of course).
+
+This PEP and its reference implementation therefore use the support for
+annotation scope. Note that this usage is a separable part from the
+implementation of :pep:`649` and :pep:`695` which provides a somewhat similar
+deferred execution model for annotations. Instead it's up to the tag function to
+evaluate any interpolations.
+
+With annotation scope in place, lambda-wrapped expressions in interpolations
+then provide the usual lexical scoping seen with f-strings. So there's no need
+to use ``locals()``, ``globals()``, or frame introspection with
+``sys._getframe`` to evaluate the interpolation. In addition, the code of each
+expression is available and does not have to be looked up with
+``inspect.getsource`` or some other means.
+
+Format Specification
+--------------------
+
+The ``format_spec`` is by default ``None`` if it is not specified in the tag string's
+corresponding interpolation.
+
+Because the tag function is completely responsible for processing ``Decoded``
+and ``Interpolation`` values, there is no required interpretation for the format
+spec and conversion in an interpolation. For example, this is a valid usage:
+
+.. code-block:: python
+
+    html'<div id={id:int}>{content:HTML|str}</div>'
```

Comment:
I know this is just an example, and I understand that might be done on  purpose, but isn't `id=` supposed to be in quotes in html ? `<id="...">` ? I understand that here  `int` may return a quoted string, but this might not be obvious.

In general I'm worried of showing slightly inaccurate examples as users tend to copy-past.

Reply:
It's a good point -- both your comment and this example. 😉 It's popular in JSX to do as shown here. In fact, my IDE defaults to curly braces, no quotes. The renderer adds the quotes in. Quotes with a brace inside (as you've described) also works.

There are other syntaxes like this. `<input checked={is_checked}>` collapses to `<input checked>`. The most well-known is spread props: `<MyComponent {...props}>`.

If you'd like, we can point out both syntaxes are legal. But that's less about tag strings and DSLs in general, and more about a particular implementation of a particular language. Or: just drop making the point illustrating tag strings doing transformations.

Reply:
My feeling is that such syntaxes are quite helpful when targetting a DSL while being Pythonic. Tag functions can of course choose what they support. Two alternatives that come to mind that could work well with tag strings:

*  `<{MyComponent(props)}>...<//>` - `MyComponent` and `props` are just usual lexically scoped Python names in this expression. `<//>` simplifies writing the closing tag (vs `</{MyComponent(props)}>, although the tag function gets to decide what satisfies a closing tag, so it could still be `</{MyComponent}>`.

* `<MyComponent {props}>`...`</MyComponent>` requires that the tag function has been configured to use a registry for `MyComponent`; `props` here is lexically scoped, and if it's like a dict, it can be expanded accordingly.

Reply:
Thanks @Carreau for the comment here.

I'm going to resolve this conversation, cause it's needed to merge the PEP. I agree that leaving this as-is is okay, since it's what JSX does, which a lot of JavaScript people have experience with. If more people speak up about this being confusing, we can revisit.